stack/dispatcher.go

```diff
@@ -69,8 +69,14 @@ func (d *Dispatcher) CheckTx(ctx basecoin.Context, store state.KVStore, tx basec
 	if err != nil {
 		return res, err
 	}
-	// TODO: check on callback
-	cb := d
+
+	// make sure no monkey business with the context
+	cb := secureCheck(d, ctx)
+
+	// and isolate the permissions and the data store for this app
+	ctx = withApp(ctx, r.Name())
+	store = stateSpace(store, r.Name())
```

Comment:
If you see how an app is created... it looks like this (from roles/middleware_test.go):

```
	disp := stack.NewDispatcher(
		stack.WrapHandler(roles.NewHandler()),
		stack.WrapHandler(stack.CheckHandler{}),
	)
	// and wrap with the roles middleware
	app := stack.New(roles.NewMiddleware()).Use(disp)
```

We create the stack, by adding all handlers to the dispatcher and the middleware to the stack.  The stack package handles the wiring and does it properly.

Clearly you can build the binary another way, the point is that we allow you to run each module runs in it's sandbox. In fact, we make it the easiest way to use the software.